ch_pipeline/processing/daily.py

```diff
@@ -407,29 +410,136 @@ def _load_hook(self):
+            < end_time,  # which contain data that includes start time and end time
+            di.CorrFileInfo.finish_time >= start_time,
+            di.ArchiveFileCopy.node == online_node,  # that are online
+            di.ArchiveFileCopy.has_file == "Y",
+        )
+
+        # chimestack files available on nearline, which include between start and end_time
+        files_nearline = archive_files.where(
+            di.ArchiveInst.name == "chimestack",
+            di.CorrFileInfo.start_time < end_time,
+            di.CorrFileInfo.finish_time >= start_time,
+            di.ArchiveFileCopy.node == nearline_node,  # that are on nearline
+            di.ArchiveFileCopy.has_file == "Y",
+        )
+
+        # dict: {acqName/fileName: absolute path}
+        filenames_online = {
+            os.path.join(f.file.acq.name, f.file.name): os.path.join(
+                f.node.root, f.file.acq.name, f.file.name
+            )
+            for f in files_online
+        }
+
+        # contains all of the files which are on nearline, but not online
+
+        # list: [absolute path]
+
+        # we need acq_name/file_name to identify a file in different locations
+        filenames_not_online = [
+            os.path.join(f.node.root, f.file.acq.name, f.file.name)
+            for f in files_nearline
+            if os.path.join(f.file.acq.name, f.file.name) not in filenames_online
+        ]
+
+        if not filenames_not_online:
+            return list(filenames_online.values())
+        else:
+            # if there are files available on nearline, which are not available online
+            # we want to return an empty list
+            return []
+
     def _available_tags(self):
         """Return all the tags that are available to run.
 
         This includes any that currently exist or are in the job queue.
         """
 
-        # TODO: should decide availability based on what data is actually
-        # available
-        # - Need to find all correlator data in the range
-        # - Figure out which ones are on cedar
-        # - Figure out which sidereal days are covered by this range
-
         csds = []
 
         # For each interval find and add all CSDs that have not already been added
         for interval in self._intervals:
-            csd_i = csds_in_range(*interval)
+            csd_i = csds_in_range(*interval, step=1)
+            csd_set = set(csds)
+            csds += [csd for csd in csd_i if csd not in csd_set]
+
+        # For each intervals_weekly, find and add all CSDs, skipping every 7
+        for interval in self._intervals_weekly:
+            csd_i = csds_in_range(*interval, step=7)
             csd_set = set(csds)
             csds += [csd for csd in csd_i if csd not in csd_set]
 
-        tags = ["%i" % csd for csd in csds]
+        # only queue jobs for which all data is available in chime_online
+        tags = ["%i" % csd for csd in csds if self._available_files(csd, csd + 1)]
```

Comment:
I think this isn't the best way to do this. It's going to fire off 4 separate database queries for every CSD in the set specified by intervals. That's far too many queries.

A more efficient way to do this is to do one set of queries and return all files that exist anywhere, and all files that exist on online (you can probably restrict to between the min and max CSDs in your list), and then count the number of files in each category that overlap with the CSD. Only if those numbers are equal is the CSD "available".

Reply:
> You should fetch the start and end times for each file when you do the queries. (remember to specify them in the .select(...)).
> And then you just calculate which files overlap. Given that none of the files overlap with each other, and none of the days overlap with each other, there's probably an efficient way of computing the file-day overlaps that doesn't involve doing all pairs.